File: espresso.rb
class Espresso
  # Accessor for Access Var
  attr_accessor :key, :iv, :counter, :ls
    def initialize(key, iv)
      # Init Variable
      @key = key
      @iv = iv
      @counter = 0
      @ls = Array.new()

      (0..1999).each do |i|
        @ls.push("x")
      end

      # Insert Key
      @key = @key.map {|x| 
      ("%08b" % x).chars
      }

      @key.each_index {|y| 
      @key[y].each {|z| 
      @ls.insert((y * 8), z.to_i)
        }
      }

      # Insert IV
      @iv = @iv.map {|x| 
      ("%08b" % x).chars
      }

      @iv.each_index {|y| 
      @iv[y].each {|z| 
      @ls.insert(128 + (y * 8), z.to_i)
        }
      }

      # Insert 1's
      (0..30).each do |i|
        @ls.insert((128 + 96 + i), 1)
      end

      # End 256 bit Init Register with 0
      @ls.insert(255, 0)

      # Run Update Init 256x
      (0..255).each do |i|
        update(true)
      end

    end

    def update(init)
      # Run Register
      out  = @ls[80 + @counter] ^ @ls[99 + @counter] ^ @ls[137 + @counter] ^ @ls[227 + @counter] ^ @ls[222 + @counter] ^ @ls[187 + @counter] ^ @ls[243 + @counter] & @ls[217 + @counter] ^ @ls[247 + @counter] & @ls[231 + @counter] ^ @ls[213 + @counter] & @ls[235 + @counter] ^ @ls[255 + @counter] & @ls[251 + @counter] ^ @ls[181 + @counter] & @ls[239 + @counter] ^ @ls[174 + @counter] & @ls[44 + @counter]  ^  @ls[164 + @counter] & @ls[29 + @counter]  ^ @ls[255 + @counter] & @ls[247 + @counter] & @ls[243 + @counter] & @ls[213 + @counter] & @ls[181 + @counter] & @ls[174 + @counter]
      n255 = @ls[0 + @counter] ^ @ls[41 + @counter] & @ls[70 + @counter]
      n251 = @ls[42 + @counter] & @ls[83 + @counter]  ^ @ls[8 + @counter]
      n247 = @ls[44 + @counter] & @ls[102 + @counter] ^ @ls[40 + @counter]
      n243 = @ls[43 + @counter] & @ls[118 + @counter] ^ @ls[103 + @counter]
      n239 = @ls[46 + @counter] & @ls[141 + @counter] ^ @ls[117 + @counter]
      n235 = @ls[67 + @counter] & @ls[90 + @counter] & @ls[110 + @counter] & @ls[137 + @counter]
      n231 = @ls[50 + @counter] & @ls[159 + @counter] ^ @ls[189 + @counter]
      n217 = @ls[3 + @counter] & @ls[32 + @counter]
      n213 = @ls[4 + @counter] & @ls[45 + @counter]
      n209 = @ls[6 + @counter] & @ls[64 + @counter]
      n205 = @ls[5 + @counter] & @ls[80 + @counter]
      n201 = @ls[8 + @counter] & @ls[103 + @counter]
      n197 = @ls[29 + @counter] & @ls[52 + @counter] & @ls[72 + @counter] & @ls[99 + @counter]
      n193 = @ls[12 + @counter] & @ls[121 + @counter]

      if(init)
          n255 ^= out
          n217 ^= out
      end

      # Update State
      @counter += 1

      @ls[255 + @counter] = n255
      @ls[251 + @counter] ^= n251
      @ls[247 + @counter] ^= n247
      @ls[243 + @counter] ^= n243
      @ls[239 + @counter] ^= n239
      @ls[235 + @counter] ^= n235
      @ls[231 + @counter] ^= n231
      @ls[217 + @counter] ^= n217
      @ls[213 + @counter] ^= n213
      @ls[209 + @counter] ^= n209
      @ls[205 + @counter] ^= n205
      @ls[201 + @counter] ^= n201
      @ls[197 + @counter] ^= n197
      @ls[193 + @counter] ^= n193

      # ??? (Copied from Original Test Vector, Maybe for Dynamic Array?)
      if(@counter == 1700)
          # Some memcpy (memcpy(ls, ls+1700, 256);)
          @counter = 0
      end

      return out
    end

end

# Set Key and IV
key = []
iv = []
  
(0..15).each do |i|
  key.push(i)
end
  
(0..11).each do |i|
  iv.push(i)
end

esp = Espresso.new(key, iv)
puts esp.ls

ky = []

(0..159).each do |i|
  ky.insert(i, esp.update(false))
end

keystream = ""
kp = ""

ky.each{
  |x| 
  kp += x.to_s

  if (kp.length == 8)
    kp = kp.reverse
    keystream += "%02x" % kp.to_i(2)
    kp = ""
  end
}

puts keystream

File: espresso.py
class espresso:
    def __init__(self, key, iv):
        # Init Variables
        self.key = key
        self.iv = iv
        self.counter = 0
        self.ls = []

        # Init Register 
        for _ in range(2000): self.ls.append(None)

        # Insert Key to Register
        for i in range(len(self.key)):
            k = list(format(key[i], '08b'))
            k.reverse()

            for j in range(len(k)): 
                self.ls[(8 * i) + j] = int(k[j])

        # Insert IV to Register
        for i in range(len(self.iv)):
            v = list(format(iv[i], '08b'))
            v.reverse()

            for j in range(len(v)): 
                self.ls[128 + (8 * i) + j] = int(v[j])
        
        # Insert "1"'s to Register 
        for i in range(31):
            self.ls[128 + 96 + i] = 1
        
        # End 256 bit Init Register with "0"
        self.ls[255] = 0

        # Init Register Run (x 256)
        for _ in range(256):
            self.update(1) 
        
        
    def update(self, init):
        # Run Register
        out  = self.ls[80 + self.counter] ^ self.ls[99 + self.counter] ^ self.ls[137 + self.counter] ^ self.ls[227 + self.counter] ^ self.ls[222 + self.counter] ^ self.ls[187 + self.counter] ^ self.ls[243 + self.counter] & self.ls[217 + self.counter] ^ self.ls[247 + self.counter] & self.ls[231 + self.counter] ^ self.ls[213 + self.counter] & self.ls[235 + self.counter] ^ self.ls[255 + self.counter] & self.ls[251 + self.counter] ^ self.ls[181 + self.counter] & self.ls[239 + self.counter] ^ self.ls[174 + self.counter] & self.ls[44 + self.counter]  ^  self.ls[164 + self.counter] & self.ls[29 + self.counter]  ^ self.ls[255 + self.counter] & self.ls[247 + self.counter] & self.ls[243 + self.counter] & self.ls[213 + self.counter] & self.ls[181 + self.counter] & self.ls[174 + self.counter]
        n255 = self.ls[0 + self.counter] ^ self.ls[41 + self.counter] & self.ls[70 + self.counter]
        n251 = self.ls[42 + self.counter] & self.ls[83 + self.counter]  ^ self.ls[8 + self.counter]
        n247 = self.ls[44 + self.counter] & self.ls[102 + self.counter] ^ self.ls[40 + self.counter]
        n243 = self.ls[43 + self.counter] & self.ls[118 + self.counter] ^ self.ls[103 + self.counter]
        n239 = self.ls[46 + self.counter] & self.ls[141 + self.counter] ^ self.ls[117 + self.counter]
        n235 = self.ls[67 + self.counter] & self.ls[90 + self.counter] & self.ls[110 + self.counter] & self.ls[137 + self.counter]
        n231 = self.ls[50 + self.counter] & self.ls[159 + self.counter] ^ self.ls[189 + self.counter]
        n217 = self.ls[3 + self.counter] & self.ls[32 + self.counter]
        n213 = self.ls[4 + self.counter] & self.ls[45 + self.counter]
        n209 = self.ls[6 + self.counter] & self.ls[64 + self.counter]
        n205 = self.ls[5 + self.counter] & self.ls[80 + self.counter]
        n201 = self.ls[8 + self.counter] & self.ls[103 + self.counter]
        n197 = self.ls[29 + self.counter] & self.ls[52 + self.counter] & self.ls[72 + self.counter] & self.ls[99 + self.counter]
        n193 = self.ls[12 + self.counter] & self.ls[121 + self.counter]

        if(init):
            n255 ^= out
            n217 ^= out
        
        # Update State
        self.counter += 1

        self.ls[255 + self.counter] = n255
        self.ls[251 + self.counter] ^= n251
        self.ls[247 + self.counter] ^= n247
        self.ls[243 + self.counter] ^= n243
        self.ls[239 + self.counter] ^= n239
        self.ls[235 + self.counter] ^= n235
        self.ls[231 + self.counter] ^= n231
        self.ls[217 + self.counter] ^= n217
        self.ls[213 + self.counter] ^= n213
        self.ls[209 + self.counter] ^= n209
        self.ls[205 + self.counter] ^= n205
        self.ls[201 + self.counter] ^= n201
        self.ls[197 + self.counter] ^= n197
        self.ls[193 + self.counter] ^= n193

        # ??? (Copied from Original Test Vector, Maybe for Dynamic Array?)
        if(self.counter == 1700):
            # Some memcpy (memcpy(ls, ls+1700, 256);)
            self.counter = 0

        return out


if __name__ == "__main__":
    # Key 128 bits / 16 Bytes
    key = [] 
    for x in range(16): key.append(int(x))

    # IV 96 bits / 12 Bytes
    iv = [] 
    for y in range(12): iv.append(int(y))

    # Keystream size (20 Bytes)
    ks = []
    for _ in range(20 * 8): ks.append(0)

    esp = espresso(key, iv)

    ky = []
    for _ in range(20 * 8): ky.append(esp.update(0))
    
    keystream = ""
    kp = ""

    # Print Keystream
    for z in range(len(ks)):
        p = ks[z] ^ ky[z]
        kp += str(p)

        if(len(kp) == 8):
            kp = ''.join(reversed(kp))
            keystream += hex(int(kp, 2))[2:]
            kp = ""
    
    print(keystream)
